ckan/tests/helpers.py

```diff
@@ -468,3 +544,149 @@ def wrapper(*args, **kwargs):
             return return_value
         return nose.tools.make_decorator(func)(wrapper)
     return decorator
+
+
+@contextlib.contextmanager
+def recorded_logs(logger=None, level=logging.DEBUG,
+                  override_disabled=True, override_global_level=True):
+    u'''
+    Context manager for recording log messages.
+
+    :param logger: The logger to record messages from. Can either be a
+        :py:class:`logging.Logger` instance or a string with the
+        logger's name. Defaults to the root logger.
+
+    :param int level: Temporary log level for the target logger while
+        the context manager is active. Pass ``None`` if you don't want
+        the level to be changed. The level is automatically reset to its
+        original value when the context manager is left.
+
+    :param bool override_disabled: A logger can be disabled by setting
+        its ``disabled`` attribute. By default, this context manager
+        sets that attribute to ``False`` at the beginning of its
+        execution and resets it when the context manager is left. Set
+        ``override_disabled`` to ``False`` to keep the current value
+        of the attribute.
+
+    :param bool override_global_level: The ``logging.disable`` function
+        allows one to install a global minimum log level that takes
+        precedence over a logger's own level. By default, this context
+        manager makes sure that the global limit is at most ``level``,
+        and reduces it if necessary during its execution. Set
+        ``override_global_level`` to ``False`` to keep the global limit.
+
+    :returns: A recording log handler that listens to ``logger`` during
+        the execution of the context manager.
+    :rtype: :py:class:`RecordingLogHandler`
+
+    Example::
+
+        import logging
+
+        logger = logging.getLogger(__name__)
+
+        with recorded_logs(logger) as logs:
+            logger.info(u'Hello, world!')
+
+        logs.assert_log(u'info', u'world')
+    '''
+    if logger is None:
+        logger = logging.getLogger()
+    elif not isinstance(logger, logging.Logger):
+        logger = logging.getLogger(logger)
+    handler = RecordingLogHandler()
+    old_level = logger.level
+    manager_level = logger.manager.disable
+    disabled = logger.disabled
+    logger.addHandler(handler)
+    try:
+        if level is not None:
+            logger.setLevel(level)
+        if override_disabled:
+            logger.disabled = False
+        if override_global_level:
+            if (level is None) and (manager_level > old_level):
+                logger.manager.disable = old_level
+            elif (level is not None) and (manager_level > level):
+                logger.manager.disable = level
+        yield handler
+    finally:
+        logger.handlers.remove(handler)
+        logger.setLevel(old_level)
+        logger.disabled = disabled
+        logger.manager.disable = manager_level
+
+
+class RecordingLogHandler(logging.Handler):
```

Comment:
`RecordingLogHandler` is intended to be used with the `recorded_logs` context manager, so it's usually only active during a short period of time. In addition you usually attach it only to the logger that you're interested in. I'm therefore not sure whether the effort of storing the logs in Redis would be worth it.

Perhaps the documentation of `RecordingLogHandler` should be updated to reflect this, in my experience it's rather useless on its own -- see all the work that `recorded_logs` has to do in order to make sure that logs are actually recorded during tests.